File: E-Commerce App Project (Tabbed)/Controllers/AccountCRUDViewController.m
//
//  AccountCRUDViewController.m
//  E-Commerce App Project (Tabbed)
//
//  Created by Rony Banik on 12/11/18.
//  Copyright © 2018 Rony Banik. All rights reserved.
//

#import "AccountCRUDViewController.h"
#import "AccountOperations.h"
#import "Validation.h"

@interface AccountCRUDViewController (){
    AccountOperations *accountOperationsObj;
    Validation *validatorObj;
    //For Edit/Update Account
    NSUserDefaults *defaults;
    NSDictionary *userData;
}

@end

@implementation AccountCRUDViewController

@synthesize firstName,lastName,email,password,confirmPassword,phone,country,state,city,postalCode,address,validationStatus;

//For Edit Account
@synthesize editAccountFirstName,editAccountLastName,editAccountEmail,editAccountPassword,editAccountConfirmPassword,editAccountPhone,editAccountCountry,editAccountState,editAccountCity,editAccountPostalCode,editAccountAddress,editAccountValidationStatus;
@synthesize registrationComplete;

- (void)viewDidLoad {
    [super viewDidLoad];
    validationStatus.hidden=YES;
    accountOperationsObj = [[AccountOperations alloc]init];
    validatorObj = [[Validation alloc]init];
    
    
    [firstName ajw_attachValidator:[validatorObj requiredMinLengthValidator:@"First Name is Required!" IntegerforMinLength:1 minLengthErrorMessage:@"It has to be something at least!" withLabelForValidationRules:validationStatus]];
    [editAccountFirstName ajw_attachValidator:[validatorObj requiredMinLengthValidator:@"First Name is Required!" IntegerforMinLength:1 minLengthErrorMessage:@"It has to be something at least!" withLabelForValidationRules:validationStatus]];
    
    [lastName ajw_attachValidator:[validatorObj requiredMinLengthValidator:@"Last Name is Required!" IntegerforMinLength:1 minLengthErrorMessage:@"Last Name Please!" withLabelForValidationRules:validationStatus]];
    [editAccountLastName ajw_attachValidator:[validatorObj requiredMinLengthValidator:@"Last Name is Required!" IntegerforMinLength:1 minLengthErrorMessage:@"Last Name Please!" withLabelForValidationRules:validationStatus]];
    
    [phone ajw_attachValidator:[validatorObj phoneValidator:validationStatus]];
    [editAccountPhone ajw_attachValidator:[validatorObj phoneValidator:validationStatus]];
    
    [email ajw_attachValidator:[validatorObj emailValidator:validationStatus]];
    [editAccountEmail ajw_attachValidator:[validatorObj emailValidator:validationStatus]];
    
    [password ajw_attachValidator:[validatorObj requiredMinLengthValidator:@"Password is required!" IntegerforMinLength:6 minLengthErrorMessage:@"It must be at least 6 charecters!" withLabelForValidationRules:validationStatus]];
    
    [editAccountPassword ajw_attachValidator:[validatorObj requiredMinLengthValidator:@"Password is required!" IntegerforMinLength:6 minLengthErrorMessage:@"It must be at least 6 charecters!" withLabelForValidationRules:validationStatus]];
    [confirmPassword ajw_attachValidator:[validatorObj requiredValidator:@"Required and Should be same as 'Password'" withLabelForValidationRules:validationStatus]];
    [editAccountConfirmPassword ajw_attachValidator:[validatorObj requiredValidator:@"Required and Should be same as 'Password'" withLabelForValidationRules:validationStatus]];
    [country ajw_attachValidator:[validatorObj requiredValidator:@"Required!" withLabelForValidationRules:validationStatus]];
    [state ajw_attachValidator:[validatorObj requiredValidator:@"Required!" withLabelForValidationRules:validationStatus]];
    [city ajw_attachValidator:[validatorObj requiredValidator:@"Required!" withLabelForValidationRules:validationStatus]];
    [postalCode ajw_attachValidator:[validatorObj requiredValidator:@"Required!" withLabelForValidationRules:validationStatus]];
    [address ajw_attachValidator:[validatorObj requiredValidator:@"Required!" withLabelForValidationRules:validationStatus]];
    [editAccountCountry ajw_attachValidator:[validatorObj requiredValidator:@"Required!" withLabelForValidationRules:validationStatus]];
    [editAccountState ajw_attachValidator:[validatorObj requiredValidator:@"Required!" withLabelForValidationRules:validationStatus]];
    [editAccountCity ajw_attachValidator:[validatorObj requiredValidator:@"Required!" withLabelForValidationRules:validationStatus]];
    [editAccountPostalCode ajw_attachValidator:[validatorObj requiredValidator:@"Required!" withLabelForValidationRules:validationStatus]];
    [editAccountAddress ajw_attachValidator:[validatorObj requiredValidator:@"Required!" withLabelForValidationRules:validationStatus]];
    
        defaults = [NSUserDefaults standardUserDefaults];
        if (![defaults boolForKey:@"SeesionUserLoggedIN"]) {
            NSLog(@"user not logged IN");
        }else{
            NSLog(@"%@",[defaults valueForKey:@"LoggedInUsersDetail"]);
            userData = [defaults dictionaryForKey:@"LoggedInUsersDetail"];
            
            editAccountFirstName.text=[userData valueForKey:@"firstName"];
            editAccountLastName.text=[userData valueForKey:@"lastName"];
            editAccountEmail.text=[userData valueForKey:@"usersEmail"];
            editAccountPhone.text=[userData valueForKey:@"phone"];
            editAccountCountry.text=[userData valueForKey:@"country"];
            editAccountState.text=[userData valueForKey:@"state"];
            editAccountCity.text=[userData valueForKey:@"city"];
            editAccountPostalCode.text=[userData valueForKey:@"postalCode"];
            editAccountAddress.text=[userData valueForKey:@"address"];
        }
    
    
    registrationComplete = [UIAlertController
                      alertControllerWithTitle:@"Registered!"
                      message:@"You've Successfully Registered!"
                      preferredStyle:UIAlertControllerStyleAlert];
    
    // Do any additional setup after loading the view.
    UIAlertAction* yesButton = [UIAlertAction
                                actionWithTitle:@"Log In!"
                                style:UIAlertActionStyleDefault
                                handler:^(UIAlertAction * action) {
                                    //Handle your yes please button action here
                                    [self.navigationController popToRootViewControllerAnimated:YES];
                                }];
    
    [registrationComplete addAction:yesButton];
    
}




/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

-(void)registrationAction:(id)sender{
    NSString *encryptedPassword = [accountOperationsObj sha1:password.text];
    NSString *encryptedConfirmedPassword = [accountOperationsObj sha1:confirmPassword.text];
    NSString *secureKeyForServerAccess = @"sdfsdfsd38792F423F4528482B4D6250655368566D597133743677397A24432646294A40kjsdhfkjsdhf";
    NSDictionary *dataToSend = [NSDictionary dictionaryWithObjectsAndKeys:
                                secureKeyForServerAccess,@"secureKeyForServerAccess_Enabled",
                                @"REGISTER_USER",@"actionRequest",firstName.text, @"firstName",
                                lastName.text, @"lastName",email.text,@"email",
                                encryptedPassword,@"password",
                                encryptedConfirmedPassword,@"confirmPassword",
                                phone.text,@"phone",country.text,@"country",
                                state.text,@"state",city.text,@"city",
                                postalCode.text,@"postalCode",address.text,@"address",
                                nil];
    [accountOperationsObj sendRequestToServer:dataToSend callback:^(NSError *error, BOOL success, NSString *customErrorMessage){
        if (success) {
            if([customErrorMessage isEqualToString:@"Registraition Successful"]){
                [sender setTitle:@"Registered" forState:UIControlStateNormal];
                [sender setBackgroundColor:[UIColor blueColor]];
                [self presentViewController:self->registrationComplete animated:YES completion:nil];
            }else{
                [self generalAlerts:@"Alert!" withMessage:customErrorMessage withYesActionTitle:@"Cancel Registration!" withNoActionTitle:@"Fix it!"];
            }
        }else{
            [self generalAlerts:@"Alert!" withMessage:customErrorMessage withYesActionTitle:@"Cancel Registration!" withNoActionTitle:@"Fix it!"];
            [sender setTitle:@"Try Again" forState:UIControlStateNormal];
        }
    }];
}

- (IBAction)registerAccount:(id)sender {
    //firstName,lastName,email,password,confirmPassword,phone,country,state,city,postalCode,address;
    if ([firstName hasText]&&[lastName hasText]&&[email hasText]&&[password hasText]&&[confirmPassword hasText]&&[phone hasText]&&[country hasText]&&[state hasText]&&[city hasText]&&[postalCode hasText]&&[address hasText]) {
            if ([accountOperationsObj validateEmailAccount:email.text]) {
                if ([password.text isEqualToString:confirmPassword.text]) {
                    [self registrationAction:sender];
                }else{
                    [self generalAlerts:@"Password Mismatch!" withMessage:@"Confirm password and password Must be same" withYesActionTitle:@"Log In!" withNoActionTitle:@"Back To Form!"];
                }
            }else{
                [self generalAlerts:@"Invalid Email!" withMessage:@"Please Insert a valid email address" withYesActionTitle:@"Cancel Registration!" withNoActionTitle:@"Ok!"];
            }
    }else{
        [self generalAlerts:@"Empty Form" withMessage:@"Please complete the form!" withYesActionTitle:@"Already Registered!" withNoActionTitle:@"Got it!"];
    }
}
-(void)editAction:(id)sender{
    NSString *encryptedPassword = [accountOperationsObj sha1:editAccountPassword.text];
    NSString *encryptedConfirmedPassword = [accountOperationsObj sha1:editAccountConfirmPassword.text];
    NSString *secureKeyForServerAccess = @"sdfsdfsd38792F423F4528482B4D6250655368566D597133743677397A24432646294A40kjsdhfkjsdhf";
    NSDictionary *dataToSend = [NSDictionary dictionaryWithObjectsAndKeys:
                                secureKeyForServerAccess,@"secureKeyForServerAccess_Enabled",
                                @"EDIT_USER",@"actionRequest",editAccountFirstName.text, @"firstName",
                                editAccountLastName.text, @"lastName",[userData valueForKey:@"usersEmail"],@"email",
                                encryptedPassword,@"password",
                                encryptedConfirmedPassword,@"confirmPassword",
                                editAccountPhone.text,@"phone",editAccountCity.text,@"country",
                                editAccountState.text,@"state",editAccountCity.text,@"city",
                                editAccountPostalCode.text,@"postalCode",editAccountAddress.text,@"address",
                                nil];
    [accountOperationsObj sendRequestToServer:dataToSend callback:^(NSError *error, BOOL success, NSString* customErrorMessage){
        if (success) {
            if([customErrorMessage isEqualToString:@"Update Successful"]){
                [sender setTitle:@"Account Updated!" forState:UIControlStateNormal];
                [sender setBackgroundColor:[UIColor blueColor]];
                [self generalAlerts:@"Account Updated!" withMessage:@"Your Account Have Successfully Updated!" withYesActionTitle:@"Log In!" withNoActionTitle:@"OK"];
                [self presentViewController:self->registrationComplete animated:YES completion:nil];
            }else{
                [self generalAlerts:@"Alert!" withMessage:customErrorMessage withYesActionTitle:@"Cancel Updating!" withNoActionTitle:@"Fix it!"];
            }
        }else{
            [self generalAlerts:@"Alert!" withMessage:customErrorMessage withYesActionTitle:@"Cancel Updating!" withNoActionTitle:@"Fix it!"];
            [sender setTitle:@"Try Again" forState:UIControlStateNormal];
            [sender setBackgroundColor:[UIColor redColor]];
        }
    }];
}
- (IBAction)editAccount:(id)sender {
    //firstName,lastName,email,password,confirmPassword,phone,country,state,city,postalCode,address;
    if ([editAccountFirstName hasText]&&[editAccountLastName hasText]&&[editAccountEmail hasText]&&[editAccountPassword hasText]&&[editAccountConfirmPassword hasText]&&[editAccountPhone hasText]&&[editAccountCountry hasText]&&[editAccountState hasText]&&[editAccountCity hasText]&&[editAccountPostalCode hasText]&&[editAccountAddress hasText]) {
        
            if ([accountOperationsObj validateEmailAccount:editAccountEmail.text]) {
                if ([editAccountPassword.text isEqualToString:editAccountConfirmPassword.text]) {
                    [self editAction:sender];
                }else{
                    [self generalAlerts:@"Password Mismatch!" withMessage:@"Confirm password and password Must be same" withYesActionTitle:@"Cancel Updating!" withNoActionTitle:@"Back To Form!"];
                }
            }else{
                [self generalAlerts:@"Invalid Email!" withMessage:@"Please Insert a valid email address" withYesActionTitle:@"Cancel Updating!" withNoActionTitle:@"Ok!"];
            }
        
    }else{
        [self generalAlerts:@"Empty Form" withMessage:@"Please complete the form!" withYesActionTitle:@"Cancel Updating!" withNoActionTitle:@"Got it!"];
    }
}
-(void)generalAlerts:(NSString *)alertTitle
         withMessage:(NSString *)alertMessage
  withYesActionTitle:(NSString *)yesActionTitle
   withNoActionTitle:(NSString *)noActionTitle{
    
    UIAlertController *generalAlert;
    generalAlert = [UIAlertController
                            alertControllerWithTitle:alertTitle
                            message:alertMessage
                            preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction* yesAction = [UIAlertAction
                                actionWithTitle:yesActionTitle
                                style:UIAlertActionStyleDefault
                                handler:^(UIAlertAction * action) {
                                    //Handle your yes please button action here
                                    [self.navigationController popToRootViewControllerAnimated:YES];
                                }];
    UIAlertAction* noAction = [UIAlertAction
                                   actionWithTitle:noActionTitle
                                   style:UIAlertActionStyleDefault
                                   handler:^(UIAlertAction * action) {
                                       
                                   }];
    [generalAlert addAction:yesAction];
    [generalAlert addAction:noAction];
    [self presentViewController:generalAlert animated:YES completion:nil];
}
@end

File: E-Commerce App Project (Tabbed)/Controllers/AccountCRUDViewController.swift
//
//  AccountCRUDViewController.swift
//  E-Commerce App Project (Tabbed)
//
//  Converted to Swift
//

import UIKit

class AccountCRUDViewController: UIViewController {
    
    @IBOutlet var firstName: UITextField!
    @IBOutlet var lastName: UITextField!
    @IBOutlet var email: UITextField!
    @IBOutlet var password: UITextField!
    @IBOutlet var confirmPassword: UITextField!
    @IBOutlet var phone: UITextField!
    @IBOutlet var country: UITextField!
    @IBOutlet var state: UITextField!
    @IBOutlet var city: UITextField!
    @IBOutlet var postalCode: UITextField!
    @IBOutlet var address: UITextField!
    @IBOutlet var validationStatus: UILabel!
    
    @IBOutlet var editAccountFirstName: UITextField!
    @IBOutlet var editAccountLastName: UITextField!
    @IBOutlet var editAccountEmail: UITextField!
    @IBOutlet var editAccountPassword: UITextField!
    @IBOutlet var editAccountConfirmPassword: UITextField!
    @IBOutlet var editAccountPhone: UITextField!
    @IBOutlet var editAccountCountry: UITextField!
    @IBOutlet var editAccountState: UITextField!
    @IBOutlet var editAccountCity: UITextField!
    @IBOutlet var editAccountPostalCode: UITextField!
    @IBOutlet var editAccountAddress: UITextField!
    @IBOutlet var editAccountValidationStatus: UILabel!
    
    var registrationComplete: UIAlertController!
    
    private var accountOperationsObj: AccountOperations!
    private var validatorObj: Validation!
    private var defaults: UserDefaults!
    private var userData: [String: Any]?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        validationStatus.isHidden = true
        accountOperationsObj = AccountOperations()
        validatorObj = Validation()
        
        firstName.ajw_attachValidator(validatorObj.requiredMinLengthValidator("First Name is Required!", integerforMinLength: 1, minLengthErrorMessage: "It has to be something at least!", withLabelForValidationRules: validationStatus))
        editAccountFirstName.ajw_attachValidator(validatorObj.requiredMinLengthValidator("First Name is Required!", integerforMinLength: 1, minLengthErrorMessage: "It has to be something at least!", withLabelForValidationRules: validationStatus))
        
        lastName.ajw_attachValidator(validatorObj.requiredMinLengthValidator("Last Name is Required!", integerforMinLength: 1, minLengthErrorMessage: "Last Name Please!", withLabelForValidationRules: validationStatus))
        editAccountLastName.ajw_attachValidator(validatorObj.requiredMinLengthValidator("Last Name is Required!", integerforMinLength: 1, minLengthErrorMessage: "Last Name Please!", withLabelForValidationRules: validationStatus))
        
        phone.ajw_attachValidator(validatorObj.phoneValidator(validationStatus))
        editAccountPhone.ajw_attachValidator(validatorObj.phoneValidator(validationStatus))
        
        email.ajw_attachValidator(validatorObj.emailValidator(validationStatus))
        editAccountEmail.ajw_attachValidator(validatorObj.emailValidator(validationStatus))
        
        password.ajw_attachValidator(validatorObj.requiredMinLengthValidator("Password is required!", integerforMinLength: 6, minLengthErrorMessage: "It must be at least 6 charecters!", withLabelForValidationRules: validationStatus))
        editAccountPassword.ajw_attachValidator(validatorObj.requiredMinLengthValidator("Password is required!", integerforMinLength: 6, minLengthErrorMessage: "It must be at least 6 charecters!", withLabelForValidationRules: validationStatus))
        
        confirmPassword.ajw_attachValidator(validatorObj.requiredValidator("Required and Should be same as 'Password'", withLabelForValidationRules: validationStatus))
        editAccountConfirmPassword.ajw_attachValidator(validatorObj.requiredValidator("Required and Should be same as 'Password'", withLabelForValidationRules: validationStatus))
        
        country.ajw_attachValidator(validatorObj.requiredValidator("Required!", withLabelForValidationRules: validationStatus))
        state.ajw_attachValidator(validatorObj.requiredValidator("Required!", withLabelForValidationRules: validationStatus))
        city.ajw_attachValidator(validatorObj.requiredValidator("Required!", withLabelForValidationRules: validationStatus))
        postalCode.ajw_attachValidator(validatorObj.requiredValidator("Required!", withLabelForValidationRules: validationStatus))
        address.ajw_attachValidator(validatorObj.requiredValidator("Required!", withLabelForValidationRules: validationStatus))
        
        editAccountCountry.ajw_attachValidator(validatorObj.requiredValidator("Required!", withLabelForValidationRules: validationStatus))
        editAccountState.ajw_attachValidator(validatorObj.requiredValidator("Required!", withLabelForValidationRules: validationStatus))
        editAccountCity.ajw_attachValidator(validatorObj.requiredValidator("Required!", withLabelForValidationRules: validationStatus))
        editAccountPostalCode.ajw_attachValidator(validatorObj.requiredValidator("Required!", withLabelForValidationRules: validationStatus))
        editAccountAddress.ajw_attachValidator(validatorObj.requiredValidator("Required!", withLabelForValidationRules: validationStatus))
        
        defaults = UserDefaults.standard
        if !defaults.bool(forKey: "SeesionUserLoggedIN") {
            print("user not logged IN")
        } else {
            print("\(defaults.value(forKey: "LoggedInUsersDetail") ?? "")")
            userData = defaults.dictionary(forKey: "LoggedInUsersDetail")
            
            editAccountFirstName.text = userData?["firstName"] as? String
            editAccountLastName.text = userData?["lastName"] as? String
            editAccountEmail.text = userData?["usersEmail"] as? String
            editAccountPhone.text = userData?["phone"] as? String
            editAccountCountry.text = userData?["country"] as? String
            editAccountState.text = userData?["state"] as? String
            editAccountCity.text = userData?["city"] as? String
            editAccountPostalCode.text = userData?["postalCode"] as? String
            editAccountAddress.text = userData?["address"] as? String
        }
        
        registrationComplete = UIAlertController(
            title: "Registered!",
            message: "You've Successfully Registered!",
            preferredStyle: .alert
        )
        
        let yesButton = UIAlertAction(title: "Log In!", style: .default) { [weak self] _ in
            self?.navigationController?.popToRootViewController(animated: true)
        }
        
        registrationComplete.addAction(yesButton)
    }
    
    func registrationAction(_ sender: UIButton) {
        let encryptedPassword = accountOperationsObj.sha1(password.text ?? "")
        let encryptedConfirmedPassword = accountOperationsObj.sha1(confirmPassword.text ?? "")
        let secureKeyForServerAccess = "sdfsdfsd38792F423F4528482B4D6250655368566D597133743677397A24432646294A40kjsdhfkjsdhf"
        
        let dataToSend: [String: Any] = [
            "secureKeyForServerAccess_Enabled": secureKeyForServerAccess,
            "actionRequest": "REGISTER_USER",
            "firstName": firstName.text ?? "",
            "lastName": lastName.text ?? "",
            "email": email.text ?? "",
            "password": encryptedPassword,
            "confirmPassword": encryptedConfirmedPassword,
            "phone": phone.text ?? "",
            "country": country.text ?? "",
            "state": state.text ?? "",
            "city": city.text ?? "",
            "postalCode": postalCode.text ?? "",
            "address": address.text ?? ""
        ]
        
        accountOperationsObj.sendRequest(toServer: dataToSend) { [weak self] error, success, customErrorMessage in
            guard let self = self else { return }
            
            if success {
                if customErrorMessage == "Registraition Successful" {
                    sender.setTitle("Registered", for: .normal)
                    sender.backgroundColor = .blue
                    self.present(self.registrationComplete, animated: true, completion: nil)
                } else {
                    self.generalAlerts("Alert!", withMessage: customErrorMessage ?? "", withYesActionTitle: "Cancel Registration!", withNoActionTitle: "Fix it!")
                }
            } else {
                self.generalAlerts("Alert!", withMessage: customErrorMessage ?? "", withYesActionTitle: "Cancel Registration!", withNoActionTitle: "Fix it!")
                sender.setTitle("Try Again", for: .normal)
            }
        }
    }
    
    @IBAction func registerAccount(_ sender: UIButton) {
        if firstName.hasText && lastName.hasText && email.hasText && password.hasText && confirmPassword.hasText && phone.hasText && country.hasText && state.hasText && city.hasText && postalCode.hasText && address.hasText {
            if accountOperationsObj.validateEmailAccount(email.text ?? "") {
                if password.text == confirmPassword.text {
                    registrationAction(sender)
                } else {
                    generalAlerts("Password Mismatch!", withMessage: "Confirm password and password Must be same", withYesActionTitle: "Log In!", withNoActionTitle: "Back To Form!")
                }
            } else {
                generalAlerts("Invalid Email!", withMessage: "Please Insert a valid email address", withYesActionTitle: "Cancel Registration!", withNoActionTitle: "Ok!")
            }
        } else {
            generalAlerts("Empty Form", withMessage: "Please complete the form!", withYesActionTitle: "Already Registered!", withNoActionTitle: "Got it!")
        }
    }
    
    func editAction(_ sender: UIButton) {
        let encryptedPassword = accountOperationsObj.sha1(editAccountPassword.text ?? "")
        let encryptedConfirmedPassword = accountOperationsObj.sha1(editAccountConfirmPassword.text ?? "")
        let secureKeyForServerAccess = "sdfsdfsd38792F423F4528482B4D6250655368566D597133743677397A24432646294A40kjsdhfkjsdhf"
        
        let dataToSend: [String: Any] = [
            "secureKeyForServerAccess_Enabled": secureKeyForServerAccess,
            "actionRequest": "EDIT_USER",
            "firstName": editAccountFirstName.text ?? "",
            "lastName": editAccountLastName.text ?? "",
            "email": userData?["usersEmail"] as? String ?? "",
            "password": encryptedPassword,
            "confirmPassword": encryptedConfirmedPassword,
            "phone": editAccountPhone.text ?? "",
            "country": editAccountCity.text ?? "",
            "state": editAccountState.text ?? "",
            "city": editAccountCity.text ?? "",
            "postalCode": editAccountPostalCode.text ?? "",
            "address": editAccountAddress.text ?? ""
        ]
        
        accountOperationsObj.sendRequest(toServer: dataToSend) { [weak self] error, success, customErrorMessage in
            guard let self = self else { return }
            
            if success {
                if customErrorMessage == "Update Successful" {
                    sender.setTitle("Account Updated!", for: .normal)
                    sender.backgroundColor = .blue
                    self.generalAlerts("Account Updated!", withMessage: "Your Account Have Successfully Updated!", withYesActionTitle: "Log In!", withNoActionTitle: "OK")
                    self.present(self.registrationComplete, animated: true, completion: nil)
                } else {
                    self.generalAlerts("Alert!", withMessage: customErrorMessage ?? "", withYesActionTitle: "Cancel Updating!", withNoActionTitle: "Fix it!")
                }
            } else {
                self.generalAlerts("Alert!", withMessage: customErrorMessage ?? "", withYesActionTitle: "Cancel Updating!", withNoActionTitle: "Fix it!")
                sender.setTitle("Try Again", for: .normal)
                sender.backgroundColor = .red
            }
        }
    }
    
    @IBAction func editAccount(_ sender: UIButton) {
        if editAccountFirstName.hasText && editAccountLastName.hasText && editAccountEmail.hasText && editAccountPassword.hasText && editAccountConfirmPassword.hasText && editAccountPhone.hasText && editAccountCountry.hasText && editAccountState.hasText && editAccountCity.hasText && editAccountPostalCode.hasText && editAccountAddress.hasText {
            if accountOperationsObj.validateEmailAccount(editAccountEmail.text ?? "") {
                if editAccountPassword.text == editAccountConfirmPassword.text {
                    editAction(sender)
                } else {
                    generalAlerts("Password Mismatch!", withMessage: "Confirm password and password Must be same", withYesActionTitle: "Cancel Updating!", withNoActionTitle: "Back To Form!")
                }
            } else {
                generalAlerts("Invalid Email!", withMessage: "Please Insert a valid email address", withYesActionTitle: "Cancel Updating!", withNoActionTitle: "Ok!")
            }
        } else {
            generalAlerts("Empty Form", withMessage: "Please complete the form!", withYesActionTitle: "Cancel Updating!", withNoActionTitle: "Got it!")
        }
    }
    
    func generalAlerts(_ alertTitle: String, withMessage alertMessage: String, withYesActionTitle yesActionTitle: String, withNoActionTitle noActionTitle: String) {
        let generalAlert = UIAlertController(
            title: alertTitle,
            message: alertMessage,
            preferredStyle: .alert
        )
        
        let yesAction = UIAlertAction(title: yesActionTitle, style: .default) { [weak self] _ in
            self?.navigationController?.popToRootViewController(animated: true)
        }
        
        let noAction = UIAlertAction(title: noActionTitle, style: .default, handler: nil)
        
        generalAlert.addAction(yesAction)
        generalAlert.addAction(noAction)
        present(generalAlert, animated: true, completion: nil)
    }
}
